// @trezor/protocol/lib/protocol-thp/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TAG_LENGTH = exports.CRC_LENGTH = exports.THP_DEFAULT_CHANNEL = exports.THP_CONTINUATION_PACKET = exports.THP_CONTROL_BYTE_DECRYPTED = exports.THP_CONTROL_BYTE_ENCRYPTED = exports.THP_READ_ACK_HEADER_BYTE = exports.THP_ERROR_HEADER_BYTE = exports.THP_HANDSHAKE_COMPLETION_RESPONSE = exports.THP_HANDSHAKE_COMPLETION_REQUEST = exports.THP_HANDSHAKE_INIT_RESPONSE = exports.THP_HANDSHAKE_INIT_REQUEST = exports.THP_CREATE_CHANNEL_RESPONSE = exports.THP_CREATE_CHANNEL_REQUEST = void 0;
exports.THP_CREATE_CHANNEL_REQUEST = 0x40;
exports.THP_CREATE_CHANNEL_RESPONSE = 0x41;
exports.THP_HANDSHAKE_INIT_REQUEST = 0x00;
exports.THP_HANDSHAKE_INIT_RESPONSE = 0x01;
exports.THP_HANDSHAKE_COMPLETION_REQUEST = 0x02;
exports.THP_HANDSHAKE_COMPLETION_RESPONSE = 0x03;
exports.THP_ERROR_HEADER_BYTE = 0x42;
exports.THP_READ_ACK_HEADER_BYTE = 0x20;
exports.THP_CONTROL_BYTE_ENCRYPTED = 0x04;
exports.THP_CONTROL_BYTE_DECRYPTED = 0x05;
exports.THP_CONTINUATION_PACKET = 0x80;
exports.THP_DEFAULT_CHANNEL = Buffer.from([0xff, 0xff]);
exports.CRC_LENGTH = 4;
exports.TAG_LENGTH = 16;
