// @solflare-wallet/metamask-sdk/lib/cjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addSignature = exports.serializeTransactionMessage = exports.serializeTransaction = exports.isLegacyTransactionInstance = void 0;
function isLegacyTransactionInstance(transaction) {
    return transaction.version === undefined;
}
exports.isLegacyTransactionInstance = isLegacyTransactionInstance;
function serializeTransaction(transaction) {
    return isLegacyTransactionInstance(transaction)
        ? transaction.serialize({
            verifySignatures: false,
            requireAllSignatures: false
        })
        : transaction.serialize();
}
exports.serializeTransaction = serializeTransaction;
function serializeTransactionMessage(transaction) {
    return isLegacyTransactionInstance(transaction)
        ? transaction.serializeMessage()
        : transaction.message.serialize();
}
exports.serializeTransactionMessage = serializeTransactionMessage;
function addSignature(transaction, publicKey, signature) {
    if (isLegacyTransactionInstance(transaction)) {
        transaction.addSignature(publicKey, Buffer.from(signature));
    }
    else {
        const signerPubkeys = transaction.message.staticAccountKeys.slice(0, transaction.message.header.numRequiredSignatures);
        const signerIndex = signerPubkeys.findIndex((pubkey) => pubkey.equals(publicKey));
        if (signerIndex >= 0) {
            transaction.signatures[signerIndex] = signature;
        }
    }
}
exports.addSignature = addSignature;
