// @trezor/connect-common/lib/map-releases.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.firmwareAssets = void 0;
exports.firmwareAssets = {
    release: {
        'releases.v1': require('../files/firmware/release/releases.v1.json'),
    },
    t1b1: {
        bitcoinonly: {
            't1b1-1.10.0-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.10.0-bitcoinonly.json'),
            't1b1-1.10.1-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.10.1-bitcoinonly.json'),
            't1b1-1.10.2-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.10.2-bitcoinonly.json'),
            't1b1-1.10.3-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.10.3-bitcoinonly.json'),
            't1b1-1.10.4-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.10.4-bitcoinonly.json'),
            't1b1-1.10.5-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.10.5-bitcoinonly.json'),
            't1b1-1.11.1-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.11.1-bitcoinonly.json'),
            't1b1-1.11.2-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.11.2-bitcoinonly.json'),
            't1b1-1.12.1-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.12.1-bitcoinonly.json'),
            't1b1-1.13.0-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.13.0-bitcoinonly.json'),
            't1b1-1.13.1-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.13.1-bitcoinonly.json'),
            't1b1-1.8.3-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.8.3-bitcoinonly.json'),
            't1b1-1.9.0-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.9.0-bitcoinonly.json'),
            't1b1-1.9.1-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.9.1-bitcoinonly.json'),
            't1b1-1.9.2-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.9.2-bitcoinonly.json'),
            't1b1-1.9.3-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.9.3-bitcoinonly.json'),
            't1b1-1.9.4-bitcoinonly': require('../files/firmware/t1b1/bitcoinonly/t1b1-1.9.4-bitcoinonly.json'),
        },
        universal: {
            't1b1-1.10.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.10.0-universal.json'),
            't1b1-1.10.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.10.1-universal.json'),
            't1b1-1.10.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.10.2-universal.json'),
            't1b1-1.10.3-universal': require('../files/firmware/t1b1/universal/t1b1-1.10.3-universal.json'),
            't1b1-1.10.4-universal': require('../files/firmware/t1b1/universal/t1b1-1.10.4-universal.json'),
            't1b1-1.10.5-universal': require('../files/firmware/t1b1/universal/t1b1-1.10.5-universal.json'),
            't1b1-1.11.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.11.1-universal.json'),
            't1b1-1.11.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.11.2-universal.json'),
            't1b1-1.12.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.12.1-universal.json'),
            't1b1-1.13.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.13.0-universal.json'),
            't1b1-1.13.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.13.1-universal.json'),
            't1b1-1.3.6-universal': require('../files/firmware/t1b1/universal/t1b1-1.3.6-universal.json'),
            't1b1-1.4.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.4.0-universal.json'),
            't1b1-1.4.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.4.1-universal.json'),
            't1b1-1.4.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.4.2-universal.json'),
            't1b1-1.5.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.5.0-universal.json'),
            't1b1-1.5.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.5.1-universal.json'),
            't1b1-1.5.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.5.2-universal.json'),
            't1b1-1.6.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.6.0-universal.json'),
            't1b1-1.6.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.6.1-universal.json'),
            't1b1-1.6.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.6.2-universal.json'),
            't1b1-1.6.3-universal': require('../files/firmware/t1b1/universal/t1b1-1.6.3-universal.json'),
            't1b1-1.7.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.7.1-universal.json'),
            't1b1-1.7.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.7.2-universal.json'),
            't1b1-1.7.3-universal': require('../files/firmware/t1b1/universal/t1b1-1.7.3-universal.json'),
            't1b1-1.8.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.8.0-universal.json'),
            't1b1-1.8.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.8.1-universal.json'),
            't1b1-1.8.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.8.2-universal.json'),
            't1b1-1.8.3-universal': require('../files/firmware/t1b1/universal/t1b1-1.8.3-universal.json'),
            't1b1-1.9.0-universal': require('../files/firmware/t1b1/universal/t1b1-1.9.0-universal.json'),
            't1b1-1.9.1-universal': require('../files/firmware/t1b1/universal/t1b1-1.9.1-universal.json'),
            't1b1-1.9.2-universal': require('../files/firmware/t1b1/universal/t1b1-1.9.2-universal.json'),
            't1b1-1.9.3-universal': require('../files/firmware/t1b1/universal/t1b1-1.9.3-universal.json'),
            't1b1-1.9.4-universal': require('../files/firmware/t1b1/universal/t1b1-1.9.4-universal.json'),
        },
    },
    t2b1: {
        bitcoinonly: {
            't2b1-2.6.3-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.6.3-bitcoinonly.json'),
            't2b1-2.6.4-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.6.4-bitcoinonly.json'),
            't2b1-2.7.0-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.7.0-bitcoinonly.json'),
            't2b1-2.7.2-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.7.2-bitcoinonly.json'),
            't2b1-2.8.0-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.8.0-bitcoinonly.json'),
            't2b1-2.8.10-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.8.10-bitcoinonly.json'),
            't2b1-2.8.7-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.8.7-bitcoinonly.json'),
            't2b1-2.8.9-bitcoinonly': require('../files/firmware/t2b1/bitcoinonly/t2b1-2.8.9-bitcoinonly.json'),
        },
        universal: {
            't2b1-2.6.3-universal': require('../files/firmware/t2b1/universal/t2b1-2.6.3-universal.json'),
            't2b1-2.6.4-universal': require('../files/firmware/t2b1/universal/t2b1-2.6.4-universal.json'),
            't2b1-2.7.0-universal': require('../files/firmware/t2b1/universal/t2b1-2.7.0-universal.json'),
            't2b1-2.7.2-universal': require('../files/firmware/t2b1/universal/t2b1-2.7.2-universal.json'),
            't2b1-2.8.0-universal': require('../files/firmware/t2b1/universal/t2b1-2.8.0-universal.json'),
            't2b1-2.8.10-universal': require('../files/firmware/t2b1/universal/t2b1-2.8.10-universal.json'),
            't2b1-2.8.7-universal': require('../files/firmware/t2b1/universal/t2b1-2.8.7-universal.json'),
            't2b1-2.8.9-universal': require('../files/firmware/t2b1/universal/t2b1-2.8.9-universal.json'),
        },
    },
    t2t1: {
        bitcoinonly: {
            't2t1-2.1.5-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.1.5-bitcoinonly.json'),
            't2t1-2.1.6-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.1.6-bitcoinonly.json'),
            't2t1-2.1.7-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.1.7-bitcoinonly.json'),
            't2t1-2.1.8-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.1.8-bitcoinonly.json'),
            't2t1-2.3.0-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.0-bitcoinonly.json'),
            't2t1-2.3.1-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.1-bitcoinonly.json'),
            't2t1-2.3.2-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.2-bitcoinonly.json'),
            't2t1-2.3.3-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.3-bitcoinonly.json'),
            't2t1-2.3.4-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.4-bitcoinonly.json'),
            't2t1-2.3.5-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.5-bitcoinonly.json'),
            't2t1-2.3.6-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.3.6-bitcoinonly.json'),
            't2t1-2.4.0-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.4.0-bitcoinonly.json'),
            't2t1-2.4.1-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.4.1-bitcoinonly.json'),
            't2t1-2.4.2-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.4.2-bitcoinonly.json'),
            't2t1-2.4.3-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.4.3-bitcoinonly.json'),
            't2t1-2.5.1-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.5.1-bitcoinonly.json'),
            't2t1-2.5.2-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.5.2-bitcoinonly.json'),
            't2t1-2.5.3-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.5.3-bitcoinonly.json'),
            't2t1-2.6.0-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.6.0-bitcoinonly.json'),
            't2t1-2.6.3-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.6.3-bitcoinonly.json'),
            't2t1-2.6.4-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.6.4-bitcoinonly.json'),
            't2t1-2.7.0-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.7.0-bitcoinonly.json'),
            't2t1-2.7.2-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.7.2-bitcoinonly.json'),
            't2t1-2.8.1-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.8.1-bitcoinonly.json'),
            't2t1-2.8.10-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.8.10-bitcoinonly.json'),
            't2t1-2.8.7-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.8.7-bitcoinonly.json'),
            't2t1-2.8.8-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.8.8-bitcoinonly.json'),
            't2t1-2.8.9-bitcoinonly': require('../files/firmware/t2t1/bitcoinonly/t2t1-2.8.9-bitcoinonly.json'),
        },
        universal: {
            't2t1-2.0.10-universal': require('../files/firmware/t2t1/universal/t2t1-2.0.10-universal.json'),
            't2t1-2.0.5-universal': require('../files/firmware/t2t1/universal/t2t1-2.0.5-universal.json'),
            't2t1-2.0.6-universal': require('../files/firmware/t2t1/universal/t2t1-2.0.6-universal.json'),
            't2t1-2.0.7-universal': require('../files/firmware/t2t1/universal/t2t1-2.0.7-universal.json'),
            't2t1-2.0.8-universal': require('../files/firmware/t2t1/universal/t2t1-2.0.8-universal.json'),
            't2t1-2.0.9-universal': require('../files/firmware/t2t1/universal/t2t1-2.0.9-universal.json'),
            't2t1-2.1.0-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.0-universal.json'),
            't2t1-2.1.1-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.1-universal.json'),
            't2t1-2.1.4-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.4-universal.json'),
            't2t1-2.1.5-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.5-universal.json'),
            't2t1-2.1.6-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.6-universal.json'),
            't2t1-2.1.7-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.7-universal.json'),
            't2t1-2.1.8-universal': require('../files/firmware/t2t1/universal/t2t1-2.1.8-universal.json'),
            't2t1-2.3.0-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.0-universal.json'),
            't2t1-2.3.1-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.1-universal.json'),
            't2t1-2.3.2-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.2-universal.json'),
            't2t1-2.3.3-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.3-universal.json'),
            't2t1-2.3.4-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.4-universal.json'),
            't2t1-2.3.5-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.5-universal.json'),
            't2t1-2.3.6-universal': require('../files/firmware/t2t1/universal/t2t1-2.3.6-universal.json'),
            't2t1-2.4.0-universal': require('../files/firmware/t2t1/universal/t2t1-2.4.0-universal.json'),
            't2t1-2.4.1-universal': require('../files/firmware/t2t1/universal/t2t1-2.4.1-universal.json'),
            't2t1-2.4.2-universal': require('../files/firmware/t2t1/universal/t2t1-2.4.2-universal.json'),
            't2t1-2.4.3-universal': require('../files/firmware/t2t1/universal/t2t1-2.4.3-universal.json'),
            't2t1-2.5.1-universal': require('../files/firmware/t2t1/universal/t2t1-2.5.1-universal.json'),
            't2t1-2.5.2-universal': require('../files/firmware/t2t1/universal/t2t1-2.5.2-universal.json'),
            't2t1-2.5.3-universal': require('../files/firmware/t2t1/universal/t2t1-2.5.3-universal.json'),
            't2t1-2.6.0-universal': require('../files/firmware/t2t1/universal/t2t1-2.6.0-universal.json'),
            't2t1-2.6.3-universal': require('../files/firmware/t2t1/universal/t2t1-2.6.3-universal.json'),
            't2t1-2.6.4-universal': require('../files/firmware/t2t1/universal/t2t1-2.6.4-universal.json'),
            't2t1-2.7.0-universal': require('../files/firmware/t2t1/universal/t2t1-2.7.0-universal.json'),
            't2t1-2.7.2-universal': require('../files/firmware/t2t1/universal/t2t1-2.7.2-universal.json'),
            't2t1-2.8.1-universal': require('../files/firmware/t2t1/universal/t2t1-2.8.1-universal.json'),
            't2t1-2.8.10-universal': require('../files/firmware/t2t1/universal/t2t1-2.8.10-universal.json'),
            't2t1-2.8.7-universal': require('../files/firmware/t2t1/universal/t2t1-2.8.7-universal.json'),
            't2t1-2.8.8-universal': require('../files/firmware/t2t1/universal/t2t1-2.8.8-universal.json'),
            't2t1-2.8.9-universal': require('../files/firmware/t2t1/universal/t2t1-2.8.9-universal.json'),
        },
    },
    t3b1: {
        bitcoinonly: {
            't3b1-2.8.10-bitcoinonly': require('../files/firmware/t3b1/bitcoinonly/t3b1-2.8.10-bitcoinonly.json'),
            't3b1-2.8.3-bitcoinonly': require('../files/firmware/t3b1/bitcoinonly/t3b1-2.8.3-bitcoinonly.json'),
            't3b1-2.8.7-bitcoinonly': require('../files/firmware/t3b1/bitcoinonly/t3b1-2.8.7-bitcoinonly.json'),
            't3b1-2.8.9-bitcoinonly': require('../files/firmware/t3b1/bitcoinonly/t3b1-2.8.9-bitcoinonly.json'),
        },
        universal: {
            't3b1-2.8.10-universal': require('../files/firmware/t3b1/universal/t3b1-2.8.10-universal.json'),
            't3b1-2.8.3-universal': require('../files/firmware/t3b1/universal/t3b1-2.8.3-universal.json'),
            't3b1-2.8.7-universal': require('../files/firmware/t3b1/universal/t3b1-2.8.7-universal.json'),
            't3b1-2.8.9-universal': require('../files/firmware/t3b1/universal/t3b1-2.8.9-universal.json'),
        },
    },
    t3t1: {
        bitcoinonly: {
            't3t1-2.7.2-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.7.2-bitcoinonly.json'),
            't3t1-2.8.0-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.8.0-bitcoinonly.json'),
            't3t1-2.8.1-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.8.1-bitcoinonly.json'),
            't3t1-2.8.10-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.8.10-bitcoinonly.json'),
            't3t1-2.8.3-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.8.3-bitcoinonly.json'),
            't3t1-2.8.7-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.8.7-bitcoinonly.json'),
            't3t1-2.8.9-bitcoinonly': require('../files/firmware/t3t1/bitcoinonly/t3t1-2.8.9-bitcoinonly.json'),
        },
        universal: {
            't3t1-2.7.2-universal': require('../files/firmware/t3t1/universal/t3t1-2.7.2-universal.json'),
            't3t1-2.8.0-universal': require('../files/firmware/t3t1/universal/t3t1-2.8.0-universal.json'),
            't3t1-2.8.1-universal': require('../files/firmware/t3t1/universal/t3t1-2.8.1-universal.json'),
            't3t1-2.8.10-universal': require('../files/firmware/t3t1/universal/t3t1-2.8.10-universal.json'),
            't3t1-2.8.3-universal': require('../files/firmware/t3t1/universal/t3t1-2.8.3-universal.json'),
            't3t1-2.8.7-universal': require('../files/firmware/t3t1/universal/t3t1-2.8.7-universal.json'),
            't3t1-2.8.9-universal': require('../files/firmware/t3t1/universal/t3t1-2.8.9-universal.json'),
        },
    },
};
