// xrpl/dist/npm/models/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeMPTokenMetadata = exports.decodeMPTokenMetadata = exports.validateMPTokenMetadata = exports.parseTransactionFlags = exports.convertTxFlagsToNumber = exports.setTransactionFlagsToNumber = exports.parseAccountRootFlags = exports.LedgerEntry = void 0;
exports.LedgerEntry = __importStar(require("./ledger"));
var flags_1 = require("./utils/flags");
Object.defineProperty(exports, "parseAccountRootFlags", { enumerable: true, get: function () { return flags_1.parseAccountRootFlags; } });
Object.defineProperty(exports, "setTransactionFlagsToNumber", { enumerable: true, get: function () { return flags_1.setTransactionFlagsToNumber; } });
Object.defineProperty(exports, "convertTxFlagsToNumber", { enumerable: true, get: function () { return flags_1.convertTxFlagsToNumber; } });
Object.defineProperty(exports, "parseTransactionFlags", { enumerable: true, get: function () { return flags_1.parseTransactionFlags; } });
var mptokenMetadata_1 = require("./utils/mptokenMetadata");
Object.defineProperty(exports, "validateMPTokenMetadata", { enumerable: true, get: function () { return mptokenMetadata_1.validateMPTokenMetadata; } });
Object.defineProperty(exports, "decodeMPTokenMetadata", { enumerable: true, get: function () { return mptokenMetadata_1.decodeMPTokenMetadata; } });
Object.defineProperty(exports, "encodeMPTokenMetadata", { enumerable: true, get: function () { return mptokenMetadata_1.encodeMPTokenMetadata; } });
__exportStar(require("./methods"), exports);
__exportStar(require("./transactions"), exports);
__exportStar(require("./common"), exports);
